File: ros_ws/src/Data_Drone/stem_times.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.io import savemat

simid_o = 505202503
simid_p = 505202502

ORIGINAL_FILE = f"src/Data_Drone/data_files/computation_times_{simid_o}_seq.npy"
PARALLEL_FILE = f"src/Data_Drone/data_files/computation_times_{simid_p}_parallel.npy"

START_IDX = 0

original_times = np.load(ORIGINAL_FILE, allow_pickle=True)
parallel_times = np.load(PARALLEL_FILE, allow_pickle=True)

original_times = original_times.item()
original_times = np.array(original_times['comp_times'])

parallel_times = parallel_times.item()
parallel_times = np.array(parallel_times['comp_times'])

timesteps = np.arange(1, len(parallel_times) + 1)

avg_original = np.mean(original_times[START_IDX:])
avg_parallel = np.mean(parallel_times[START_IDX:])

# Plotting
plt.figure(figsize=(10, 5))
plt.plot(timesteps[START_IDX:], original_times[START_IDX:], marker='o', color='tab:blue', label="Sequential", markersize=4, linewidth=1, alpha=1)
plt.plot(timesteps[START_IDX:], parallel_times[START_IDX:], marker='o', color='tab:orange', label="Parallel", markersize=4, linewidth=1, alpha=1)

plt.axhline(avg_original, color='magenta', linestyle='--', linewidth=2, label=f"Seq Avg = {avg_original:.3f} s")
plt.axhline(avg_parallel, color='green', linestyle='--', linewidth=2, label=f"Par Avg = {avg_parallel:.3f} s")

plt.xlabel("Time Index n")
plt.ylabel("Computation Time (s)")
plt.title("MPC Solve Time per Step for 3 Drones")
plt.grid(True)
plt.legend(loc='upper right')
plt.tight_layout()

# plt.savefig("computation_time_3_drone", dpi=300, bbox_inches='tight')

plt.show()
# savemat(f"computation_times_{simid_o}_{simid_p}.mat", {
#     "seq": original_times,
#     "par": parallel_times,
# })

File: ros_ws/src/Data_Drone/matlab_scripts/stem_times.m
% Load the computation time data
sim_o = 505202504;
sim_p = 505202505;
file_name = sprintf('mats/computation_times_%d_%d.mat', sim_o, sim_p);
load(file_name);

start_idx = 1;
seq_times = seq(start_idx : end);
par_times = par(start_idx : end);


avg_seq = mean(seq_times);
avg_par = mean(par_times);

timesteps = 1:length(par_times);

figure;
hold on;
stem(timesteps(start_idx:end), seq_times(start_idx:end), 'filled', 'DisplayName', 'Sequential');
stem(timesteps(start_idx:end), par_times(start_idx:end), 'filled', 'DisplayName', 'Parallel');

yline(avg_seq, '--r', sprintf('Seq Avg = %.4f s', avg_seq), 'LabelHorizontalAlignment', 'left', 'LabelVerticalAlignment', 'middle');
yline(avg_par, '--g', sprintf('Par Avg = %.4f s', avg_par), 'LabelHorizontalAlignment', 'left', 'LabelVerticalAlignment', 'middle');

xlabel('Timestep');
ylabel('Computation Time (s)');
title('MPC Solve Time per Step');
legend;
grid on;
hold off;
